File: backend/app/data_contracts/canonical_item.py
"""Canonical item, base item, and implicit contracts."""

from __future__ import annotations

from dataclasses import dataclass, field

from .canonical_base import CanonicalRecord
from .canonical_modifier import CanonicalModifierReference


@dataclass(frozen=True)
class CanonicalImplicit(CanonicalRecord):
    modifier_references: tuple[CanonicalModifierReference, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "modifier_references", tuple(self.modifier_references))


@dataclass(frozen=True)
class CanonicalItemBase(CanonicalRecord):
    item_type: str | None = None
    slot: str | None = None
    subtype: str | None = None
    level_requirement: int | None = None
    class_restrictions: tuple[str, ...] = field(default_factory=tuple)
    implicit_ids: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "class_restrictions", tuple(self.class_restrictions))
        object.__setattr__(self, "implicit_ids", tuple(self.implicit_ids))

File: frontend/src/types/canonicalItem.ts
import type { CanonicalRecord } from "./canonicalBase";
import type { CanonicalModifierReference } from "./canonicalModifier";

export interface CanonicalImplicit extends CanonicalRecord {
  item_base_id?: string | null;
  item_type?: string | null;
  modifier_references: CanonicalModifierReference[];
  modifier_rows?: Array<Record<string, unknown>>;
  value_range?: {
    min?: number | null;
    max?: number | null;
    value_scale?: string | null;
    polarity?: string | null;
  };
  stable_calculable?: boolean;
}

export interface CanonicalItemBase extends CanonicalRecord {
  item_category?: string | null;
  item_type?: string | null;
  equipment_slot?: string | null;
  slot?: string | null;
  subtype?: string | null;
  classification?: "weapon" | "armor" | "accessory" | "idol" | "unknown" | string;
  level_requirement?: number | null;
  attribute_requirements?: Record<string, number>;
  requirements?: Record<string, unknown>;
  class_restrictions: string[];
  mastery_restrictions?: string[];
  implicit_ids: string[];
  base_stats?: Record<string, unknown>;
  stable_calculable?: boolean;
}
